### PR TITLE
Remove optimizer context

### PR DESCRIPTION
**Pull request by jdahm**
## Technical Description

Removes the OptimizerContext object. `OptimizerContext.{cpp,h}` became `Lowering.{cpp,h}`, and the salient parts were named `toStencilInstantiation` and `restoreIIR`. The latter is only used in the IIR builder.

Other changes:
- Passes no longer take a context, `PassManager` has a default constructor
- Hardware options were moved from the, now gone, `OptimizerContext` to `Optimizer/Options.inc`
- Test logic is simplified because they do not have to hold an `OptimizerContext`

## Testing

CTests are updated by this PR.

### CODE REVIEW

**Comment by jdahm** [comment]
launch jenkins

**Comment by jdahm** [comment]
launch jenkins

**Comment by jdahm** [comment]
launch jenkins

**Comment by Stagno** [comment]
@jdahm could you provide some guidance on how to review it? If you write a list of things that changed (e.g. where the logic of the OptimizerContext was moved to) we can focus on those points. thx!

**Comment by jdahm** [comment]
@Stagno no problem. here it is:

`OptimizerContext.{cpp,h} became `Lowering.{cpp,h}`, and the salient parts were named `toStencilInstantiation` and `restoreIIR`. The latter is only used in the IIR builder.

Other changes:
- Passes no longer take a context, `PassManager` has a default constructor
- Hardware options were moved from the, now gone, `OptimizerContext` to `Optimizer/Options.inc`
- Test logic is simplified because they do not have to hold an `OptimizerContext`

**Comment by jdahm** [comment]
@Stagno @mroethlin Let me know if you would like to walk through the code together. I'd like to get this merged soon, as most other PRs will conflict with it. Once it's merged we can go ahead and split the projects (finally!)

File: dawn/src/dawn/Optimizer/Driver.cpp [hunk 1]
```diff
@@ -24,4 +66,255 @@ std::list<PassGroup> defaultPassGroups() {
...
+    if(!passManager.runAllPassesOnStencilInstantiation(instantiation, options))
+      throw std::runtime_error("An error occurred.");
+
+    DAWN_LOG(INFO) << "Done with optimization and analysis passes for `" << instantiation->getName()
+                   << "`";
+
+    if(options.SerializeIIR) {
+      const IIRSerializer::Format serializationKind =
+          options.SerializeIIR ? IIRSerializer::parseFormatString(options.IIRFormat)
+                               : IIRSerializer::Format::Json;
```
Comment (Stagno) [inline dawn/src/dawn/Optimizer/Driver.cpp:276]: Have we ever used a serialization format different from json?

File: dawn/src/dawn/Optimizer/Lowering.cpp [hunk 2]
```diff
@@ -652,9 +633,11 @@ bool OptimizerContext::fillIIRFromSIR(
...
       stencilInstantiation.get(), stencilInstantiation->getIIR()->getChildren());
 
-  if(getOptions().ReportAccesses) {
-    stencilInstantiation->reportAccesses(dawn::log::info.stream());
-  }
+  // Options are not available here
+  // TODO Move this elsewhere or delete
+  // if(getOptions().ReportAccesses) {
+  //   stencilInstantiation->reportAccesses(dawn::log::info.stream());
+  // }
```
Comment (Stagno) [inline dawn/src/dawn/Optimizer/Lowering.cpp:640]: Should this be solved within this PR?

Comment (Stagno) [reply dawn/src/dawn/Optimizer/Lowering.cpp:640]: Btw there is an `options` object available, should be the one you need, right?

File: dawn/src/dawn/Optimizer/Lowering.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+struct SIR;
+namespace sir {
+struct Stencil;
+}
+namespace iir {
+class StencilInstantiation;
+}
+
+// TODO Remove this
```
Comment (Stagno) [inline dawn/src/dawn/Optimizer/Lowering.h:33]: TODO up to date?

File: dawn/src/dawn/Optimizer/PassStageReordering.cpp [hunk 4]
```diff
@@ -23,17 +22,13 @@
...
 bool PassStageReordering::run(
-    const std::shared_ptr<iir::StencilInstantiation>& stencilInstantiation) {
+    const std::shared_ptr<iir::StencilInstantiation>& stencilInstantiation,
+    const Options& options) {
   const std::string filenameWE =
       fs::path(stencilInstantiation->getMetaData().getFileName()).filename().stem();
-  if(context_.getOptions().ReportPassStageReodering)
-    stencilInstantiation->jsonDump(filenameWE + "_before.json");
+
+  // stencilInstantiation->jsonDump(filenameWE + "_before.json");
```
Comment (Stagno) [inline dawn/src/dawn/Optimizer/PassStageReordering.cpp:31]: Are these dumps not useful anymore? (also the "_after.json" one)

File: dawn/src/dawn/Optimizer/Driver.cpp (hunk 1 above)
Comment (jdahm) [reply dawn/src/dawn/Optimizer/Driver.cpp:276]: I believe this is just defaulting to Json unless we specify the output format.

File: dawn/src/dawn/Optimizer/Lowering.cpp (hunk 2 above)
Comment (jdahm) [reply dawn/src/dawn/Optimizer/Lowering.cpp:640]: Good catch!

File: dawn/src/dawn/Optimizer/Lowering.h (hunk 3 above)
Comment (jdahm) [reply dawn/src/dawn/Optimizer/Lowering.h:33]: I updated the documentation and removed the todo. Thanks for catching!

File: dawn/src/dawn/Optimizer/PassStageReordering.cpp (hunk 4 above)
Comment (jdahm) [reply dawn/src/dawn/Optimizer/PassStageReordering.cpp:31]: I added a `write-si` option that writes these to file.

**Comment by jdahm** [comment]
launch jenkins

**Comment by jdahm** [comment]
@Stagno Sorry for the delay! It should be ready for a final review now.

File: dawn/src/dawn/Optimizer/Driver.cpp (hunk 1 above)
Comment (jdahm) [reply dawn/src/dawn/Optimizer/Driver.cpp:276]: I believe @twicki thinks we should keep the byte IO option. Thoughts @twicki?

Comment (eddie-c-davis) [reply dawn/src/dawn/Optimizer/Driver.cpp:276]: I would agree, no harm in keeping that option.

**Pull request merged**